service/matching/backlog_manager_test.go

```diff
@@ -154,3 +157,164 @@ func TestTaskWriterShutdown(t *testing.T) {
+	require.Equal(t, int64(-1), backlogMgr.taskAckManager.getAckLevel(), "should only move ack level on completion")
+	require.Equal(t, int64(1), backlogMgr.taskAckManager.getReadLevel(), "read level should be 1 since a task has been added")
+	require.Equal(t, int64(1), backlogMgr.taskAckManager.completeTask(1), "should move ack level and decrease backlog counter on completion")
+
+	require.Equal(t, int64(1), backlogMgr.taskAckManager.getAckLevel(), "should only move ack level on completion")
+	require.Equal(t, int64(0), backlogMgr.db.getApproximateBacklogCount(), "0 tasks in the backlog")
+
+}
+
+func TestApproximateBacklogCounterDecrement_MultipleTasks(t *testing.T) {
+	controller := gomock.NewController(t)
+	backlogMgr := newBacklogMgr(controller, false)
+
+	backlogMgr.taskAckManager.addTask(int64(1))
+	backlogMgr.taskAckManager.addTask(int64(2))
+	backlogMgr.taskAckManager.addTask(int64(3))
+
+	// Manually updating the backlog size since adding tasks to the outstanding map does not increment the counter
+	backlogMgr.db.updateApproximateBacklogCount(int64(3))
+
+	require.Equal(t, int64(3), backlogMgr.db.getApproximateBacklogCount(), "1 task in the backlog")
+	require.Equal(t, int64(-1), backlogMgr.taskAckManager.getAckLevel(), "should only move ack level on completion")
+	require.Equal(t, int64(3), backlogMgr.taskAckManager.getReadLevel(), "read level should be 1 since a task has been added")
+
+	// Complete tasks
+	require.Equal(t, int64(-1), backlogMgr.taskAckManager.completeTask(2), "should not move the ack level")
+	require.Equal(t, int64(3), backlogMgr.db.getApproximateBacklogCount(), "should not decrease the backlog counter as ack level has not gone up")
+
+	require.Equal(t, int64(-1), backlogMgr.taskAckManager.completeTask(3), "should not move the ack level")
+	require.Equal(t, int64(3), backlogMgr.db.getApproximateBacklogCount(), "should not decrease the backlog counter as ack level has not gone up")
+
+	require.Equal(t, int64(3), backlogMgr.taskAckManager.completeTask(1), "should move the ack level")
+	require.Equal(t, int64(0), backlogMgr.db.getApproximateBacklogCount(), "should decrease the backlog counter to 0 as no more tasks in the backlog")
+
+}
+
+// TestAddTasksValidateBacklogCounter uses the "backlogManager methods" to add a task to the backlog.
+func TestAddSingleTaskValidateBacklogCounter(t *testing.T) {
+	controller := gomock.NewController(t)
+	backlogMgr := newBacklogMgr(controller, false)
+
+	// only start the taskWriter for now!
+	backlogMgr.taskWriter.Start()
+	task := &persistencespb.TaskInfo{
+		ExpiryTime: timestamp.TimeNowPtrUtcAddSeconds(3000),
+		CreateTime: timestamp.TimeNowPtrUtc(),
+	}
+	err := backlogMgr.SpoolTask(task)
+	require.NoError(t, err)
+	require.Equal(t, backlogMgr.db.getApproximateBacklogCount(), int64(1))
+	backlogMgr.taskWriter.Stop()
+}
+
+func TestAddTasksValidateBacklogCounter_ServiceError(t *testing.T) {
```

Comment:
@dnr added the test here where db throws a Service Error yet the counter has increased.